src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_InvalidStatement.cs

```diff
@@ -0,0 +1,187 @@
+        Right: IConversionExpression (ConversionKind.Cast, Implicit) (OperationKind.ConversionExpression, Type: Program, Constant: null)
+            ILiteralExpression (Text: null) (OperationKind.LiteralExpression, Type: null, Constant: null)
+  IBlockStatement (0 statements) (OperationKind.BlockStatement)
+";
+            VerifyOperationTreeForTest<IfStatementSyntax>(source, expectedOperationTree);
+        }
+
+        [Fact, WorkItem(17607, "https://github.com/dotnet/roslyn/issues/17607")]
+        public void InvalidIfElseStatement()
+        {
+            string source = @"
+using System;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        var x = new Program();
+        /*<bind>*/if ()
+        {
+        }
+        else if (x) x;
+        else
+/*</bind>*/    }
+}
+";
+            string expectedOperationTree = @"
+IIfStatement (OperationKind.IfStatement, IsInvalid)
+  Condition: IConversionExpression (ConversionKind.Invalid, Implicit) (OperationKind.ConversionExpression, Type: System.Boolean, IsInvalid)
+      IInvalidExpression (OperationKind.InvalidExpression, Type: ?, IsInvalid)
+  IBlockStatement (0 statements) (OperationKind.BlockStatement)
+  IIfStatement (OperationKind.IfStatement, IsInvalid)
+    Condition: IConversionExpression (ConversionKind.Invalid, Implicit) (OperationKind.ConversionExpression, Type: System.Boolean, IsInvalid)
+        ILocalReferenceExpression: x (OperationKind.LocalReferenceExpression, Type: Program)
+    IExpressionStatement (OperationKind.ExpressionStatement, IsInvalid)
+      ILocalReferenceExpression: x (OperationKind.LocalReferenceExpression, Type: Program)
+    IExpressionStatement (OperationKind.ExpressionStatement, IsInvalid)
+      IInvalidExpression (OperationKind.InvalidExpression, Type: ?, IsInvalid)
+";
+            VerifyOperationTreeForTest<IfStatementSyntax>(source, expectedOperationTree);
+        }
+
+        [Fact, WorkItem(17607, "https://github.com/dotnet/roslyn/issues/17607")]
+        public void InvalidForStatement()
+        {
+            string source = @"
+using System;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        var x = new Program();
+        /*<bind>*/for (P; x; )
+        {
+        }/*</bind>*/
+    }
+}
+";
+            string expectedOperationTree = @"
+IForLoopStatement (LoopKind.For) (OperationKind.LoopStatement, IsInvalid)
+  Condition: IConversionExpression (ConversionKind.Invalid, Implicit) (OperationKind.ConversionExpression, Type: System.Boolean, IsInvalid)
+      ILocalReferenceExpression: x (OperationKind.LocalReferenceExpression, Type: Program)
+  Before: IExpressionStatement (OperationKind.ExpressionStatement, IsInvalid)
+      IInvalidExpression (OperationKind.InvalidExpression, Type: ?, IsInvalid)
+  IBlockStatement (0 statements) (OperationKind.BlockStatement)
+";
+            VerifyOperationTreeForTest<ForStatementSyntax>(source, expectedOperationTree);
+        }
+    }
+}
```

Comment:
You are correct, we seem to be generating BoundBadStatement when goto/break/continue etc. fail to bind or we have an unexpected statement syntax kind (not possible from our parsers). I will add tests for the former.